linalg/solvers.cpp

```diff
@@ -188,14 +188,16 @@ void IterativeSolver::SetOperator(const Operator &op)
    }
 }
 
-void IterativeSolver::Monitor(int it, real_t norm, const Vector& r,
+bool IterativeSolver::Monitor(int it, real_t norm, const Vector& r,
                               const Vector& x, bool final) const
 {
```

Comment:
When an `IterativeSolver` with a monitor is used multiple times, we need to make sure the monitor state is reset at the start. We can automate this by adding a `Reset` method to the monitor and calling it here when `it == 0 && !final` -- the assumption is that `IterativeSolver`s will always first call the monitor with `it = 0` and `final = false`. Another option is to just add the monitor `Reset` call at the beginning of the `Mult` method of all iterative solvers.

Since it is possible for user monitor classes to add their own state that needs to be reset, the `Reset` method should be either (a) virtual or (b) call a separate virtual method (e.g. `ResetDerived`) after resetting the monitor base class state.